### ISSUE TITLE
Easy way to add an extra file for php.ini settings

### ISSUE THREAD

**Issue by stof** [heroku/heroku-buildpack-php#231]
Currently, the default memory_limit in the php.ini is 128M. While this is a good idea for PHP-FPM to allow concurrency, it is a bad default for the CLI (which should be allowed to use the whole dyno memory in a single PHP process).
Thus, overwriting the memory_limit for PHP-FPM is easy thanks to the possibility to provide a custom PHP-FPM config file, which can contain `php_admin_value[memory_limit] = 256M`. 

But overwriting the php.ini of the CLI is harder (if we don't want to force passing `-c` each time we run PHP, which we don't). And I had a use case for that actually: running `bin/console cache:warmup` in a Symfony project may require more than 128M of memory, and I wanted to have this step in the `composer compile` script to build a fully-optimized project.

My workaround for now is to use `composer compile` for that (relying on the fact that it is the first place where I can hook, but also the first place where I may need a higher memory_limit in php.ini as Composer already handles bumping the memory_limit for itself):

```json
{
    "scripts": {
        "compile": [
            "cp .tools/heroku/999-heroku.ini .heroku/php/etc/php/conf.d/",
            "npm run build-prod",
            "bin/console cache:warmup"
        ],
    }
}
```

And the `.tools/heroku/999-heroku.ini` file in my project looks like this:

```ini
[PHP]
memory_limit=512M
```

Combined with the PHP-FPM config, this allows me to configure both the CLI and the PHP-FPM memory limit.

Is it considered the best way to achieve such result ? If yes, is it worth documenting it in the buildpack documentation ? And if no, what is the best place ?

**Comment by dzuelke** [comment]
Yeah, this does come up from time to time, and I'm really torn about it. My reasoning for not having it is that if you have a magic way of setting CLI (or global) INI settings that is specific to Heroku, your code is no longer portable - if your `php makeimages.php` task or whatever requires 1024 MB of memory, then for maximum portability across environments, you should invoke it with `php -dmemory_limit=1G makeimages.php`, or do the `ini_set()` before you run that task.

Mind you you can always drop a `php-cli.ini` into the root level of your project, and PHP will automatically pick that up, but the disadvantage of course is that it fully replaces the main INI file.

**Comment by stof** [comment]
@dzuelke for local development, I'm *not* configuring the CLI php.ini with a memory_limit of 128M in the Dockerfile used by my team.
The fact that heroku does not give me easy control on this configuration is what makes me create heroku-specific logic to configure PHP (I'm fine with my existing solution for now though, as `composer run-script compile` is used only by the buildpack in my case. I don't rely on it locally)

**Comment by dzuelke** [comment]
Right but that's what I'm saying. That `Dockerfile` works, but once you take the code anywhere else (or someone changes things around), the code will stop working again, because it relied on "magic" system wide config.

Hmh... :(

**Comment by stof** [comment]
@dzuelke requiring anyone to run `php -d memory_limit=512M bin/console cache:warmup` instead of `bin/console cache:warmup` relies on "magic" manual mess up when running. I can assure you that I'm very likely to update my own system to use a higher memory limit.

Btw, for the CLI, is it a waste of resources to always limit it to 128M on Heroku unless the user does magic things to change the memory limit. It means that most of your dyno memory gets unused (until the day you notice it). It is too bad to switch to a performance dyno to figure out that your worker is still limited to 128M (i.e. 1/4 of a standard-1x dyno)

**Comment by Kocal** [comment]
It looks like you can and **must** use `@php` instead of `php`, see https://devcenter.heroku.com/articles/php-support#custom-compile-step
